src/critics/reference_trajectory_critic.cpp

```diff
@@ -11,12 +11,26 @@ void ReferenceTrajectoryCritic::initialize()
 {
   auto node = parent_.lock();
   auto getParam = utils::getParamGetter(node, name_);
-  getParam(power_, "reference_cost_power", 1);
-  getParam(weight_, "reference_cost_weight", 15.0);
+  getParam(reference_cost_power_, "reference_cost_power", 1);
+  getParam(reference_cost_weight_, "reference_cost_weight", 5.0);
+
+  getParam(enable_nearest_path_angle_critic_, "enable_nearest_path_angle_critic", true);
```

Comment:
I'm confused, aren't these covered by other critic functions that could be put in parallel? This seems to break the encapsilation of this critic as just being a reference trajectory tracking critic. I think path angle would be a reasonable critic to consume into this one since the path angle is still part of the reference trajectory, but the goal critic bit is a little weird. If you were to put the path angle critic into this one, is there a reason to keep the separate critic? 

Also, why not just have both critics going (path + path angle) vs having it merged into a single critic plugin? Keeping them modular seems more ideal

Reply:
GoalCritic and PathAngleCritic consider last goal of considered path (that's why i proposed not to use getMaxCostmapDist as max_robot_pose_search_dist_. This value could be too far away, and behavior of whole optimizer will be dependent on that. 

I merged new critics (nearest_path_angle/nearest_goal) into referenece_trajectory because there was no much choice. reference critic find for every trajectory point it's nearest path segment (which is quite expensive). This nearest segments is needed for this new critics, because instead of using last path point as a goal/goal angle, they use farthest path point among those nearest (with some offset).

Reply:
> that's why i proposed not to use getMaxCostmapDist as max_robot_pose_search_dist_

But that's not how it works, `max_robot_pose_search_dist_` is used to find the range of points to find the robot pose within on the path: https://github.com/artofnothingness/mppic/blob/edf87e1b451a721eb68a393943912a774857ce2c/src/path_handler.cpp#L33-L36

The range of path points to use from that starting point isn't parameterized: https://github.com/artofnothingness/mppic/blob/edf87e1b451a721eb68a393943912a774857ce2c/src/path_handler.cpp#L47-L53 its using the costmap bounds strictly. So I don't think that's having the impact you think it is. The only impact in changing the default of `max_robot_pose_search_dist_` is widening the window to look for the robot starting point, but it doesn't change the actual pruned path (assuming it still finds the right starting point).

> This value could be too far away, and behavior of whole optimizer will be dependent on that.

So this is implementing similar behavior, but instead of the last point, its taking something a bit closer to home? Why not instead we just add a parameter for that maximum forward to take from the starting pose found to prune the path to so instead of statically being the `costmap` bounds, its a parameterized value and then the `path angle` / `goal` critics will use that value that's closer to the robot? 

That would give you the behavior you're looking for without requiring duplication in the path reference critic, no?

Reply:
added prune_distance